2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -105,16 +115,32 @@ tx_deactivate(validator_address)
   if (state == candidate) then
     //set status to inactive at n + pipeline_length
     validators[validator_address].state[cur_epoch+pipeline_length] = inactive
+    remove_validator_from_sets(validator_address, offset)
 }
 ```
 
 ```go
 tx_reactivate(validator_address)
 {
   var state = read_epoched_field(validators[validator_address].state, cur_epoch+pipeline_length, ⊥)
-  if (state == inactive) then
+  if (state == inactive && !is_jailed(validator_address)) then
     //set status to candidate at n + pipeline_length
     validators[validator_address].state[cur_epoch+pipeline_length] = candidate
+    //add validator to the inactive set
+    add_validator_to_sets(validator_address, pipeline_length)
+}
+```
+
+```go
+tx_unjail(validator_address)
+{
+  //check validator is jailed and can be unjailed
+  if (is_jailed(validator_address) && (cur_epoch - validators[validator_address].jail_record.epoch > min_sentence)) then
```

Comment:
I think in here @sergio-mena suggested we use `cur_epoch - validators[validator_address].jail_record.epoch + pipeline_length > min_sentence)`, because we're adding the validator back to the validator sets at pipeline offset